### PR TITLE
Add `panicking::panic_count` to `mc-sgx-panic`

### PR DESCRIPTION
**Pull request by nick-mobilecoin**

### CODE REVIEW

**Comment by nick-mobilecoin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #14** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/14" target="_blank">Graphite</a> 
    * **PR #13** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/13" target="_blank">Graphite</a> 
      * **PR #15** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/15" target="_blank">Graphite</a> 
        * **PR #22** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/22" target="_blank">Graphite</a> 
          * **PR #24** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/24" target="_blank">Graphite</a> 
            * **PR #30** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/30" target="_blank">Graphite</a> 
              * **PR #31** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/31" target="_blank">Graphite</a>  👈
        * **PR #21** <a href="https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/21" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/31?utm_source=stack-comment).

**Comment by nick-mobilecoin** [comment]
Moved back to a draft, with the failures on, https://github.com/mobilecoinfoundation/sgx-std/actions/runs/3887979469/jobs/6634849726, it seems pretty clear that the value is *not* being thread local
```console
---- panicking::test::decrementing_one_at_a_time stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `5`,
 right: `4`', panic/src/panicking.rs:100:9

---- panicking::test::incrementing_one_at_a_time stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `6`,
 right: `1`', panic/src/panicking.rs:88:9

---- panicking::test::initialized_to_0 stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `6`,
 right: `0`', panic/src/panicking.rs:68:9
```

**Comment by nick-mobilecoin** [comment]
Doing some debugging have the side by side of what rust std does and what is being attempted in this PR https://godbolt.org/z/r3c37Tqxh
I also removed no_std and tried to do what rust std does and got the same failure...

**Comment by nick-mobilecoin** [comment]
moved out of draft, found problem that was causing llvm-cov to fail, added explanation in code at https://github.com/mobilecoinfoundation/sgx-std/pull/31/files#diff-0ce129c061b49b04073e4988bbaca5999d4b7148812aff7489623ab6f53edad8R66

```rust
        /// By default each test runs in a separate thread thus having their own
        /// thread local copy of `LOCAL_PANIC_COUNT`. However per
        /// https://github.com/rust-lang/rust/issues/58907 when running single
        /// threaded the thread is re-used for each test execution so the value
        /// should be reset to a known value on each test. This also means that
        /// there isn't a reliable way to test that 0 is the true initial value.
```

**Review by awygle — APPROVED** [review]
LGTM

File: panic/src/panicking.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    //! - no `set_always_abort()`. The docs say this is for `libc::fork`.
+    //!   Enclaves do not support forking or spawning child threads or processes
+    //! - use the `thread_local!` macro or the `::std::thread::LocalKey` type.
+    //!   These are generic over any type that `needs_drop`, but
+    //!   `usize` does not need drop
+    use core::cell::Cell;
+
+    /// Panic count for the current thread.
+    #[thread_local]
+    static LOCAL_PANIC_COUNT: Cell<usize> = Cell::new(0);
```
Comment (jcape) [inline panic/src/panicking.rs:29]: Why not [AtomicUsize](https://doc.rust-lang.org/core/sync/atomic/struct.AtomicUsize.html)

**Review by jcape — COMMENTED** [review]
Note: https://github.com/intel/linux-sgx/issues/283

**Comment by nick-mobilecoin** [comment]
> Note: [intel/linux-sgx#283](https://github.com/intel/linux-sgx/issues/283)

I looked through the current doc, https://download.01.org/intel-sgx/sgx-linux/2.18/docs/Intel_SGX_Developer_Reference_Linux_2.18_Open_Source.pdf trying to get a feel of how the threading worked, but wasn't able to piece it together.
If I'm reading the issue correctly, for panics, it will only be an issue if threads are bound to the untrusted thread, and the untrusted side assumes each ecall is a fresh state

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -98,9 +98,7 @@ jobs:
     strategy:
       matrix:
         rust:
-          - stable
-          - beta
-          - nightly-2022-12-13
+          - nightly-2023-01-04
```
Comment (nick-mobilecoin) [inline .github/workflows/ci.yaml:101]: Had to go back to only nightly due to `#[thread_local]`

**Review by NotGyro — APPROVED** [review]
This looks good to me.

**Comment by nick-mobilecoin** [comment]
⏳ This pull request is set to merge as part of a Graphite merge job
Stack job ID: 174FRuRmqv3ul7Wrh6cz.
See details on [graphite.dev](https://app.graphite.dev/github/pr/mobilecoinfoundation/sgx-std/31)

**Pull request merged**